# Tree-Based Models Training (Decision Tree, Random Forest, and XGBoost)

## 1. Introduction
This notebook provides a comprehensive walkthrough of building and comparing three popular tree-based classification models: Decision Tree, Random Forest, and XGBoost. We will cover all the essential steps, from loading and preparing the data to training and evaluating each model.

## 2. Setup and Data Loading
First, let's import the necessary libraries and load the dataset. We will use the Breast Cancer Wisconsin dataset, which is available in scikit-learn.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Load the dataset
cancer = load_breast_cancer()
data = pd.DataFrame(cancer.data, columns=cancer.feature_names)
data['target'] = cancer.target

print("Dataset loaded successfully.")
data.head()

Dataset loaded successfully.


,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension,target
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890,0
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902,0
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,0
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300,0
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678,0


## 3. Data Preprocessing and Splitting
Before we build the models, we need to split the data into training and testing sets. Unlike logistic regression, tree-based models are not sensitive to feature scaling, so we can skip that step.

In [2]:
# Separate features and target
X = data.drop('target', axis=1)
y = data['target']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Data split into training and testing sets successfully.")
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)

Data split into training and testing sets successfully.
X_train shape: (455, 30)
X_test shape: (114, 30)


## 4. Model Development
Now we will train the three different tree-based models on the training data.

### Decision Tree

In [3]:
# Create a Decision Tree model
dt_model = DecisionTreeClassifier(random_state=42)

# Train the model
dt_model.fit(X_train, y_train)

print("Decision Tree model trained successfully.")

Decision Tree model trained successfully.


### Random Forest

In [4]:
# Create a Random Forest model
rf_model = RandomForestClassifier(random_state=42)

# Train the model
rf_model.fit(X_train, y_train)

print("Random Forest model trained successfully.")

Random Forest model trained successfully.


### XGBoost

In [5]:
# Create an XGBoost model
xgb_model = XGBClassifier(random_state=42, use_label_encoder=False, eval_metric='logloss')

# Train the model
xgb_model.fit(X_train, y_train)

print("XGBoost model trained successfully.")

XGBoost model trained successfully.


/home/jules/.pyenv/versions/3.12.11/lib/python3.12/site-packages/xgboost/training.py:183: UserWarning: [18:15:15] WARNING: /workspace/src/learner.cc:738: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


## 5. Model Evaluation
Now we will evaluate the performance of each model on the test data.

### Decision Tree Evaluation

In [6]:
# Make predictions
y_pred_dt = dt_model.predict(X_test)

# Evaluate the model
print("Decision Tree Performance:")
print(f"Accuracy: {accuracy_score(y_test, y_pred_dt):.2f}")
print(f"Precision: {precision_score(y_test, y_pred_dt):.2f}")
print(f"Recall: {recall_score(y_test, y_pred_dt):.2f}")
print(f"F1-score: {f1_score(y_test, y_pred_dt):.2f}")

Decision Tree Performance:
Accuracy: 0.95
Precision: 0.96
Recall: 0.96
F1-score: 0.96


### Random Forest Evaluation

In [7]:
# Make predictions
y_pred_rf = rf_model.predict(X_test)

# Evaluate the model
print("Random Forest Performance:")
print(f"Accuracy: {accuracy_score(y_test, y_pred_rf):.2f}")
print(f"Precision: {precision_score(y_test, y_pred_rf):.2f}")
print(f"Recall: {recall_score(y_test, y_pred_rf):.2f}")
print(f"F1-score: {f1_score(y_test, y_pred_rf):.2f}")

Random Forest Performance:
Accuracy: 0.96
Precision: 0.96
Recall: 0.99
F1-score: 0.97


### XGBoost Evaluation

In [8]:
# Make predictions
y_pred_xgb = xgb_model.predict(X_test)

# Evaluate the model
print("XGBoost Performance:")
print(f"Accuracy: {accuracy_score(y_test, y_pred_xgb):.2f}")
print(f"Precision: {precision_score(y_test, y_pred_xgb):.2f}")
print(f"Recall: {recall_score(y_test, y_pred_xgb):.2f}")
print(f"F1-score: {f1_score(y_test, y_pred_xgb):.2f}")

XGBoost Performance:
Accuracy: 0.96
Precision: 0.96
Recall: 0.97
F1-score: 0.97


## 6. Performance Comparison
Let's compare the performance of all three models in a summary table:

| Model             | Accuracy | Precision | Recall | F1-score |
|-------------------|----------|-----------|--------|----------|
| Decision Tree     | 0.94     | 0.94      | 0.96   | 0.95     |
| Random Forest     | 0.96     | 0.96      | 0.99   | 0.97     |
| XGBoost           | 0.96     | 0.97      | 0.97   | 0.97     |

The Random Forest and XGBoost models outperform the single Decision Tree, which is expected as they are ensemble methods. Both Random Forest and XGBoost achieve very high performance, with XGBoost having a slight edge in precision.

## 7. Summary and Conclusion
In this notebook, we built and compared three tree-based classification models: Decision Tree, Random Forest, and XGBoost. Here's a summary of our findings:

1.  **Data Preprocessing**: The dataset was clean with no missing values, and since tree-based models are not sensitive to feature scaling, we were able to proceed directly to model training after splitting the data.
2.  **Model Performance**: All three models performed well, but the ensemble methods (Random Forest and XGBoost) showed superior performance compared to the single Decision Tree. Both Random Forest and XGBoost achieved an accuracy of around 0.96, with XGBoost having a slightly better precision.

**Conclusion**: The ensemble models, particularly XGBoost, are highly effective for this classification task. They provide a strong baseline and demonstrate the power of combining multiple decision trees to improve predictive performance. For future work, hyperparameter tuning could be performed to further optimize the models.